cachi2/core/package_managers/metayarn.py

```diff
@@ -0,0 +1,123 @@
+import logging
+import re
+from collections import defaultdict
+from functools import partial
+from pathlib import Path
+from typing import Iterable, Optional, Union
+
+from cachi2.core.errors import PackageRejected
+from cachi2.core.models.input import PackageInput, Request
+from cachi2.core.models.output import RequestOutput
+from cachi2.core.package_managers import yarn, yarn_classic
+from cachi2.core.resolver import _merge_outputs
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+def _contains_yarn_version_trait(
+    package: PackageInput,
+    source_dir: RootedPath,
+    trait_pattern: re.Pattern[str],
+) -> Optional[re.Match[str]]:
+    text = _get_path_to_yarn_lock(package, source_dir).read_text()
+    return trait_pattern.search(text)
+
+
+_yarnv1_pattern = re.compile("yarn lockfile v1")  # See [yarn_v1_trait].
+_yarnv2_pattern = re.compile("__metadata:")  # See [yarn_v2_trait] and [yarn_v2_test_repo].
+contains_yarn_classic = partial(_contains_yarn_version_trait, trait_pattern=_yarnv1_pattern)
+
+
+_yarn_versions = {
+    "yarn_classic": contains_yarn_classic,
+    "yarn": partial(_contains_yarn_version_trait, trait_pattern=_yarnv2_pattern),
+}
+
+_yarn_processors = {
+    "yarn_classic": yarn_classic.fetch_yarn_source,
+    "yarn": yarn.fetch_yarn_source,
+}
+
+
+def _separate_packages_by_yarn_version(
+    packages: Iterable[PackageInput],
+    source_dir: RootedPath,
+) -> dict[str, Union[list[PackageInput], tuple[PackageInput, Exception]]]:
+    """Sorts packages to bins depending on which Yarn version was used.
+
+    The output dictionary contains "uncategorized" entry to capture anything
+    that could not be categorized (likely yet-unsupported versions of Yarn).
+
+    The output dictionary also contains categroy "exceptions" to accumulate
+    exceptions that occured during pre-processing of packages.
+    """
+    output = defaultdict(list)
+    for p in packages:
+        category, exception = _yarn_selector(p, source_dir)
+        if exception is None:
+            output[category].append(p)
+        else:  # This is an exceptional result.
+            # Categories would never clash, but mypy does not know that and
+            # does not want to learn.
+            output[category].append((p, exception))  # type: ignore
+    return output  # type: ignore
+
+
+def _yarn_selector(
+    package: PackageInput,
+    source_dir: RootedPath,
+) -> tuple[str, Optional[Exception]]:
+    try:
+        for yarn_version, version_matches_for in _yarn_versions.items():
+            if version_matches_for(package, source_dir):
+                return yarn_version, None
+        else:
+            return "uncategorized", None
```

Comment:
No, I have intentionally put the final return into `for`'s `else` clause, primarily to avoid confusion about intended indentation level. I guess it dis not work as intended.

Reply:
Ohh, it's just me then >_<. I think this is literally my first time seeing a `for...else` construct "in the wild"